src/main/java/com/squareup/javawriter/Types.java

```diff
@@ -0,0 +1,351 @@
+        return o instanceof GenericArrayType
+            && Objects.equal(componentType, ((GenericArrayType) o).getGenericComponentType());
+      }
+      @Override public int hashCode() {
+        return componentType.hashCode();
+      }
+      @Override public String toString() {
+        return typeToString(this);
+      }
+    };
+  }
+
+  /**
+   * Returns a type that represents an unknown type that extends {@code bound}. For example, if
+   * {@code bound} is {@code CharSequence.class}, this returns {@code ? extends CharSequence}. If
+   * {@code bound} is {@code Object.class}, this returns {@code ?}, which is shorthand for {@code
+   * ? extends Object}.
+   */
+  public static WildcardType subtypeOf(Type bound) {
+    checkNotPrimitive(bound);
+    return wildcardType(new Type[] {bound}, EMPTY_TYPE_ARRAY);
+  }
+
+  /**
+   * Returns a type that represents an unknown supertype of {@code bound}. For example, if {@code
+   * bound} is {@code String.class}, this returns {@code ? super String}.
+   */
+  public static WildcardType supertypeOf(Type bound) {
+    checkNotPrimitive(bound);
+    return wildcardType(new Type[] {Object.class}, new Type[] {bound});
+  }
+
+  private static WildcardType wildcardType(final Type[] upperBounds, final Type[] lowerBounds) {
+    return new WildcardType() {
+      @Override public Type[] getUpperBounds() {
+        return upperBounds.clone();
+      }
+      @Override public Type[] getLowerBounds() {
+        return lowerBounds.clone();
+      }
+      @Override public boolean equals(Object o) {
+        return o instanceof WildcardType
+            && Arrays.equals(upperBounds, ((WildcardType) o).getUpperBounds())
+            && Arrays.equals(lowerBounds, ((WildcardType) o).getLowerBounds());
+      }
+      @Override public int hashCode() {
+        return Arrays.hashCode(lowerBounds) ^ Arrays.hashCode(upperBounds);
+      }
+      @Override public String toString() {
+        return typeToString(this);
+      }
+    };
+  }
+
+  public static TypeVariable<?> typeVariable(final String name, final Type... bounds) {
+    checkNotNull(name);
+    for (Type bound : bounds) {
+      checkNotPrimitive(bound);
+    }
+
+    return new TypeVariable<GenericDeclaration>() {
+      @Override public Type[] getBounds() {
+        return bounds.clone();
+      }
+      @Override public String getName() {
+        return name;
+      }
+      @Override public GenericDeclaration getGenericDeclaration() {
+        throw new UnsupportedOperationException();
+      }
+      @Override public boolean equals(Object o) {
+        return o instanceof TypeVariable
+            && name.equals(((TypeVariable<?>) o).getName());
+      }
+      @Override public int hashCode() {
+        return name.hashCode();
+      }
+      @Override public String toString() {
+        return typeToString(this);
+      }
+
+      // Java 8 requires these methods. We have them to compile, but we don't exercise them.
```

Comment:
Regarding Java 7 compatibility, if this is the only place where that comes into play, it wouldn't actually be very hard to return a dynamic proxy here instead of implementing the TypeVariable interface in code. Especially since we are throwing from the new methods anyway. I did something very similar for Guava's Types class in https://github.com/google/guava/commit/f4aa25e74a9466c2f93a2147a7cf9b01850dd41f.